geomstats/backend/numpy_linalg.py

```diff
@@ -3,15 +3,86 @@
 import numpy as np
 import scipy.linalg
 
+from geomstats.backend.numpy import to_ndarray
+
+
+def exph(x):
+    eigvals, eigvecs = np.linalg.eigh(x)
+    eigvals = np.exp(eigvals)
+    eigvals = np.vectorize(np.diag, signature='(n)->(n,n)')(eigvals)
+    transp_eigvecs = np.transpose(eigvecs, axes=(0, 2, 1))
+    result = np.matmul(eigvecs, eigvals)
+    result = np.matmul(result, transp_eigvecs)
+    return result
+
 
 def expm(x):
-    return np.vectorize(
-        scipy.linalg.expm, signature='(n,m)->(n,m)')(x)
+    ndim = x.ndim
+    new_x = to_ndarray(x, to_ndim=3)
+    if (new_x - np.transpose(new_x, axes=(0, 2, 1)) == 0).all():
+        result = exph(new_x)
+    else:
+        result = np.vectorize(scipy.linalg.expm,
+                              signature='(n,m)->(n,m)')(new_x)
+
+    if ndim == 2:
+        return result[0]
+    else:
+        return result
 
 
 def logm(x):
-    return np.vectorize(
-        scipy.linalg.logm, signature='(n,m)->(n,m)')(x)
+    ndim = x.ndim
+    new_x = to_ndarray(x, to_ndim=3)
+    if (new_x - np.transpose(new_x, axes=(0, 2, 1)) == 0).all():
+        eigvals, eigvecs = np.linalg.eigh(new_x)
+        eigvals = np.log(eigvals)
+        if (eigvals > 0).all():
+            eigvals = np.vectorize(np.diag, signature='(n)->(n,n)')(eigvals)
+            transp_eigvecs = np.transpose(eigvecs, axes=(0, 2, 1))
+            result = np.matmul(eigvecs, eigvals)
+            result = np.matmul(result, transp_eigvecs)
+        else:
+            result = np.vectorize(scipy.linalg.logm,
+                                  signature='(n,m)->(n,m)')(new_x)
+    else:
+        result = np.vectorize(scipy.linalg.logm,
+                              signature='(n,m)->(n,m)')(new_x)
+
+    if ndim == 2:
+        return result[0]
```

Comment:
Nice! This is linked to a discussion about shapes that we are having with Olivier, Niklas and Stefan

Reply:
Yes I discussed with them as well. I wonder if this trick works in any case. If yes, it should be applied in many cases according to them.